src/main/java/org/apache/tools/bzip2/CBZip2InputStream.java

```diff
@@ -29,957 +29,918 @@
+        }
+
+        for (int i = minLen, vec = 0, b = base[i]; i <= maxLen; i++) {
+            final int nb = base[i + 1];
+            vec += nb - b;
+            b = nb;
+            limit[i] = vec - 1;
+            vec <<= 1;
+        }
+
+        for (int i = minLen + 1; i <= maxLen; i++) {
+            base[i] = ((limit[i - 1] + 1) << 1) - base[i];
+        }
+    }
+
+    private void makeMaps() {
+        final boolean[] inUse = this.data.inUse;
+        final byte[] seqToUnseq = this.data.seqToUnseq;
+
+        int nInUseShadow = 0;
+
+        for (int i = 0; i < 256; i++) {
+            if (inUse[i]) seqToUnseq[nInUseShadow++] = (byte) i;
+        }
+
+        this.nInUse = nInUseShadow;
+    }
+
+    public int read() throws IOException {
+        if (this.in != null) {
+            return read0();
+        } else {
+            throw new IOException("stream closed");
+        }
+    }
+
+    public int read(final byte[] dest, final int offs, final int len) throws IOException {
+        if (offs < 0) {
+            throw new IndexOutOfBoundsException("offs(" + offs + ") < 0.");
+        }
+        if (len < 0) {
+            throw new IndexOutOfBoundsException("len(" + len + ") < 0.");
+        }
+        if (offs + len > dest.length) {
+            throw new IndexOutOfBoundsException("offs(" + offs + ") + len(" + len + ") > dest.length(" + dest.length + ").");
+        }
+        if (this.in == null) {
+            throw new IOException("stream closed");
+        }
+
+        final int hi = offs + len;
+        int destOffs = offs;
+        for (int b; (destOffs < hi) && ((b = read0()) >= 0); ) {
+            dest[destOffs++] = (byte) b;
+        }
+
+        return (destOffs == offs) ? -1 : (destOffs - offs);
+    }
+
+    private int read0() throws IOException {
+        final int retChar = this.currentChar;
+
+        switch (this.currentState) {
+            case EOF:
+                return -1;
+
+            case START_BLOCK_STATE:
+                throw new IllegalStateException();
+
+            case RAND_PART_A_STATE:
+                throw new IllegalStateException();
+
+            case RAND_PART_B_STATE:
+                setupRandPartB();
+                break;
+
+            case RAND_PART_C_STATE:
+                setupRandPartC();
+                break;
+
+            case NO_RAND_PART_A_STATE:
+                throw new IllegalStateException();
+
+            case NO_RAND_PART_B_STATE:
+                setupNoRandPartB();
+                break;
+
+            case NO_RAND_PART_C_STATE:
+                setupNoRandPartC();
+                break;
+
+            default:
+                throw new IllegalStateException();
+        }
+
+        return retChar;
+    }
+
+    private void init() throws IOException {
+        if (null == in) {
+            throw new IOException("No InputStream");
+        }
+        if (in.available() == 0) {
+            throw new IOException("Empty InputStream");
+        }
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the number of conditional operators (5) used in the expression (maximum allowed 3). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://titan.maxgamer.org:9000/coding_rules#rule_key=squid%3AS1067)